File: Graphs/BFS/BFS.py
from typing import List

class Vertex:
  def __init__(self, id: int) -> None:
    self.id = id
    self.color = ""
    self.distance = 0 # distance of this vertex from the source (s)
    self.parent = None

  # Adding the hash func and eq func to allow 
  def __hash__(self):
    return hash(self.id)

  def __eq__(self, other):
    return isinstance(other, Vertex) and self.id == other.id    


class Graph:
  def __init__(self) -> None:
    # dict to store vertex : -> [ neighbor vertices ]
    self.adj_list = {}

  
  # add edge to graph from u to v, assume directed for now
  def add_edge(self, u: Vertex, v: Vertex) -> None:
    if u not in self.adj_list:
      self.adj_list[u] = []  # initialize, the vertex to have an empty list
    if v not in self.adj_list:
      self.adj_list[v] = []  # initialize, the vertex to have an empty list

    self.adj_list[u].append(v)
    # self.graph[v].append(u) # for the case of undirected , but comment for now

  def get_vertex(self, id: int) -> Vertex:
    for vertex in self.adj_list:
      if vertex.id == id:
        return vertex
    return None
  
  def get_list_for(self, u: Vertex) -> List[Vertex]:
    if u in self.adj_list:
        return self.adj_list[u]
    return []
    

from collections import deque

def BFS(graph, start_id):
  # get the source vertex
  source = graph.get_vertex(start_id)

  # for each vertext in the graph, mark
  # WHITE(unseen),
  # distance from source (inf),
  # parent (None)
  for vertex in graph:
    if vertex.id != start_id:
      vertex.color = "WHITE"
      vertex.distance = float('inf')
      vertex.parent = None

  # Mark source GRAY(seen, but not explored)
  # distance from source = 0
  # the source has no parent since it serves as the root
  source.color = "GRAY"
  source.distance = 0
  source.parent = None

  # initialize the queue with just the source vertex
  queue = deque([source])

  # while the queue isn't empty(i.e. we have some GRAY vertices)
  while queue:
    # Pop from front of Queue
    u = queue.pop()

    # Get the neighbors for the vertex
    u_neighbors = graph.get_list_for(u)

    # for all of u's neighbors ...
    for v in u_neighbors:
      # if the neighbor is unseen
      if v.color == "WHITE":
        v.color = "GRAY" # mark the neighbor as GRAY (seen but not explored)
        v.distance = u.distance + 1  ## distance is parent's distance + 1 
        v.parent = u  # parent of neighbor is u(originating vertex / predecessor)
        queue.append(v)  # add neighbor to the queue
    
    # Mark u as fully seen and explored 
    u.color = "BLACK"

File: Graphs/BFS/BFS.cpp
#include <iostream>
#include <string>
#include <unordered_map>
#include <vector>

class Vertex {
  // TODO(oSamDavis): Making everything public for now, might be nice to modularize later on
  public:

  int id;  // id of this vertex(e.g. A, B, C, D)
  
  // Color that marks the state TODO(oSamDavis): this could possibly be an enum
  // States: UNVISITED, VISITED(but not explored), VISITED AND EXPLORED
  std::string color;

  // distance of this vertex from source {s}
  int distance;

  // predecessor or parent of this Vertex
  Vertex* parent;

  Vertex(int id) {
    this->id = id;
    color = "";
    distance = 0;
    parent = nullptr;
  }

  // Overload the eq operator for use of this class as a Key in std::unordered_map
  bool operator==(const Vertex& other) const {
    return id == other.id;
  }  

};

// Implemented std::hash<T> for Vertex so I can use as Key in std::unordered_map
// See discussion: https://shorturl.at/0a5jI
// This block of code is called: Specialization of the std::hash template for the Vertex class
// Read more: https://www.geeksforgeeks.org/template-specialization-c/
namespace std {
  template <>
  struct hash<Vertex> {
    std::size_t operator()(const Vertex& vertex) const  {
      return std::hash<int>()(vertex.id);
    }
  };
}

class Graph {
  public:
  // unordered_map to store Vertex : -> [ neighbor vertices ]
  std::unordered_map<Vertex*, std::vector<Vertex*>> adj_list_;
  
  // Add edge to graph from u to v, assume directed for now
  void AddEdge(Vertex* u, Vertex* v) {
    // if the vertex u is not in the adj list ...
    if(adj_list_.find(u) == adj_list_.end()) {
      // ... start a list for that vertex u 
      adj_list_.insert({u, std::vector<Vertex*>()});
    }

    // if the vertex v is not in the adj list ...
    if(adj_list_.find(v) == adj_list_.end()) {
      // ... start a list for that vertex v
      adj_list_.insert({v, std::vector<Vertex*>()});
    }

    // add the edge v to u
    adj_list_[u].push_back(v);
    // adj_list[u].push_back(u);  // for the case of undirected, but comment for noq
  }


  Vertex* GetVertex(int id) {

    // use an iterator to loop through the keys
    for(auto it=adj_list_.begin(); it != adj_list_.end(); ++it) {

      // if the vertex's id is equal to id passed 
      if (it->first->id == id) {
        return it->first; // return vertex
      }
    }
    return nullptr;
  }

  std::vector<Vertex*> GetListFor(Vertex* u) {
    if(adj_list_.find(u) != adj_list_.end()) {
      return adj_list_[u];
    }
    return std::vector<Vertex*>();
  }
};


void BFS(Graph graph, int start_id) {

}


int main() {


  return 0;
}